File: algorithms/N/Next Greater Numerically Balanced Number/Next Greater Numerically Balanced Number.cpp
class Solution {
public:
    bool valid(int n)
    {
        vector<int> map(10,0);
        while(n)
        {
            int rem = n%10;
            map[rem]++;
            n = n/10;
        }
        for(int i=0; i<10; i++)
            if(map[i] && map[i]!=i) return false;
        return true;
    }
    
    int nextBeautifulNumber(int n) {
     
        while(true) 
        {
            ++n;
            if(valid(n))
                return n;
        }
        return 1;
    }
};

File: algorithms/N/Next Greater Numerically Balanced Number/Next Greater Numerically Balanced Number.py
class Solution:
    def nextBeautifulNumber(self, n: int) -> int:
        n_digits = len(str(n))
        
        next_max = {
            1: [1],
            2: [22],
            3: [122, 333],
            4: [1333, 4444],
            5: [14444, 22333, 55555],
            6: [122333, 224444, 666666, 155555],
            7: [1224444, 2255555, 3334444, 1666666, 7777777]
                }
        
        if n >= int(str(n_digits) * n_digits):
            n_digits += 1
            return min(next_max[n_digits])
        
        ans = float('inf')
        for num in sorted(next_max[n_digits]):      
            cands = set(permutations(str(num)))
            cands = sorted(map(lambda x: int("".join(x)), cands))
        
            loc = bisect.bisect(cands, n)
            if loc < len(cands):           
                ans = min(ans, cands[loc])
        
        return ans
